metaseq/distributed/rendezvous.py

```diff
@@ -70,4 +70,59 @@ def _error(msg):
     raise RuntimeError("Unable to perform re-rendezvous using tcpr:// method")
 
 
+STORE_BASED_BARRIER_MARKER = "store_based_barrier_key:"
+
+
+class ComplicitStore(Store):
+    def __init__(self, store: Store, world_size: int):
+        super().__init__()
+        self.store = store
+        self.world_size = world_size
+
+    def set(self, *args, **kwargs):
+        return self.store.set(*args, **kwargs)
+
+    def get(self, *args, **kwargs):
+        return self.store.get(*args, **kwargs)
+
+    def add(self, key: str, *args, **kwargs):
+        # The marker isn't always a prefix: it's sometimes prefixed with default_pg/.
+        if STORE_BASED_BARRIER_MARKER in key:
+            return self.world_size
+        return self.store.add(key, *args, **kwargs)
+
+    def compare_set(self, *args, **kwargs):
+        return self.store.compare_set(*args, **kwargs)
+
+    def delete_key(self, *args, **kwargs):
+        return self.store.delete_key(*args, **kwargs)
+
+    def num_keys(self, *args, **kwargs):
+        return self.store.num_keys(*args, **kwargs)
+
+    def set_timeout(self, *args, **kwargs):
+        return self.store.set_timeout(*args, **kwargs)
+
+    def wait(self, *args, **kwargs):
+        return self.store.wait(*args, **kwargs)
+
+
+STORE = None
+
+
+def _tcp_retry_barrierless_rendezvous_handler(
```

Comment:
Testing this now - is there retry logic here? I don't see it...

Reply:
Oh nvm, I see the tcpr string replacement below now, lol

Reply:
This function calls into the `rendezvous` function (on line 118) using a URL that starts with `tcpr://`, which should thus in turn be rerouted to the retry logic. Does that make sense?